BibBuddy/app/src/main/java/de/bibbuddy/BookFragment.java

```diff
@@ -108,6 +125,213 @@ public boolean onOptionsItemSelected(MenuItem item) {
+          retrieveBibContent();
+          writeBibFile("Download", shelfName, bibExportContent);
+        } else {
+          Toast.makeText(getContext(), R.string.storage_permission_denied,
+              Toast.LENGTH_SHORT).show();
+        }
+        break;
+
+      default:
+    }
+  }
+
+  private void createBibFile(String folderName, String fileName) {
+    try {
+      String rootPath = Environment.getExternalStorageDirectory() + "/" + folderName + "/";
+      File root = new File(rootPath);
+
+      if (!root.exists()) {
+        root.mkdirs();
+      }
+
+      File file = new File(rootPath + fileName + ".bib");
+      if (file.exists()) {
+        file.delete();
+      }
+      file.createNewFile();
+
+      FileOutputStream out = new FileOutputStream(file);
+      out.flush();
+      out.close();
+
+      Toast.makeText(getContext(), getString(R.string.exported_file_stored_in) + '\n'
+          + "/" + folderName + "/" + fileName + ".bib", Toast.LENGTH_LONG).show();
+
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  private void retrieveBibContent() {
+
+    DatabaseHelper dbHelper = new DatabaseHelper(getContext());
+    BookDao bd = new BookDao(dbHelper);
+    NoteDao nd = new NoteDao(dbHelper);
+
+    List<Long> bookIdsCurrShelf = bd.getAllBookIdsForShelf(shelfId);
+    List<Author> authorsCurrBook = new ArrayList<>();
+    List<Long> notesCurrBook = new ArrayList<>();
+    String allNotesCurrBook = "";
+    String authorNamesCurrBook = "";
+
+    //for each book in the current shelf
+    for (int i = 0; i < bookIdsCurrShelf.size(); i++) {
+      Book currBook;
+      Long currBookId = bookIdsCurrShelf.get(i);
+      authorsCurrBook = bd.getAllAuthorsForBook(currBookId);
+      notesCurrBook = nd.getAllNoteIdsForBook(currBookId);
+      currBook = bd.findById(currBookId);
+
+      /*
+      get the notes for the current book
+      and save the bib content in one string
+      */
+      for (int k = 0; k < notesCurrBook.size(); k++) {
+        String noteTextCurrBook = nd.findTextById(notesCurrBook.get(k));
+        allNotesCurrBook +=  "annote={" + noteTextCurrBook + "}," + '\n';
+      }
+
+      /*
+      get author's first and last name and include
+      the needed book data in a bib format
+      */
+      for (int u = 0; u < authorsCurrBook.size(); u++) {
+        if (authorsCurrBook.size() > 1) {
+          authorNamesCurrBook = authorNamesCurrBook + " and "
+              + authorsCurrBook.get(u).getFirstName()
+              + " " + authorsCurrBook.get(u).getLastName();
+        } else {
+          authorNamesCurrBook = authorsCurrBook.get(u).getFirstName()
+              + " " + authorsCurrBook.get(u).getLastName();
+        }
+      }
+
+      //remove the last "and" in case of more than one author
+      if (authorNamesCurrBook.contains("and")) {
+        authorNamesCurrBook = authorNamesCurrBook.substring(0,
+            authorNamesCurrBook.length() - 5);
+      }
```

Comment:
I think this section here needs a bit of rework. Basically, the if/else-statement should contain the loop, not the other way round. See suggestion:

```suggestion
      if (authorsCurrBook.size() > 1) {
        for (int u = 0; u < authorsCurrBook.size(); u++) {
          authorNamesCurrBook = authorNamesCurrBook + " and "
              + authorsCurrBook.get(u).getFirstName()
              + " " + authorsCurrBook.get(u).getLastName();
        }
      } else {
          authorNamesCurrBook = authorsCurrBook.get(u).getFirstName()
              + " " + authorsCurrBook.get(u).getLastName();
      }
```

Reply:
This also makes the "delete last and" construction unnecessary.

Reply:
Here in this suggestion again you will have always " and " at first place. E.g. if you have **authorName** and **authorFamilyName**, you will have as output **"and authorName authorFamilyName"**

Reply:
Oh, yes - thanks for pointing that out! Altered my suggestion above.

Reply:
in your second suggestion, the else statement is outside the for-loop. So it is not possible to check if there is one/ more authors.

Reply:
The else statement is only executed if we know already that there is one author only, so there is no need to check that anymore.

Reply:
The 'u' is available only in the for-loop and not outside of it. You use 'u' in the else-statement, that is outside the for-loop.
![image](https://user-images.githubusercontent.com/37459401/109400036-a4884e00-7946-11eb-82bd-941c3fa63862.png)